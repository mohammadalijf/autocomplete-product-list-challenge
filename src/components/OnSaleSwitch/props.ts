import { SwitchProps } from "@material-ui/core";

export interface IOnSaleSwitchProps {
  /**
   * change handler for checkbox value
   */
  onChange?: OnSaleSwitchChangeEvent;
  /**
   * is checkbox checked
   */
  checked?: boolean;
}

/**
 * Event handler for OnSaleSwitch when check value is changed
 * @param checked: is checkbox checked or not
 */
export type OnSaleSwitchChangeEvent = (checked: boolean) => void;

/**
 * OnSaleSwitch' props
 */
export type OnSaleSwitchProps = IOnSaleSwitchProps &
  Omit<SwitchProps, "onChange" | "checked">;
