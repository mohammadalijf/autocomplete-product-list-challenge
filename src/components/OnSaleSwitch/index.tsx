import { FormControlLabel, Switch, SwitchProps } from "@material-ui/core";
import { FunctionComponent } from "react";
import { OnSaleSwitchProps } from "./props";

const OnSaleSwitch: FunctionComponent<OnSaleSwitchProps> = ({
  onChange,
  checked,
  ...props
}) => {
  //#region handlers

  /**
   * event handler for checkbox onChange
   * @param event onChange event from input
   */
  const changeHanlder: SwitchProps["onChange"] = (event) => {
    onChange?.(event.target.checked);
  };

  //#endregion
  return (
    <FormControlLabel
      control={
        <Switch
          {...props}
          checked={checked}
          data-testid={"onSaleSwitch"}
          color="primary"
          inputProps={{ "aria-label": "show only on sales products" }}
          onChange={changeHanlder}
        />
      }
      label="Show only on sales"
    />
  );
};

export default OnSaleSwitch;
