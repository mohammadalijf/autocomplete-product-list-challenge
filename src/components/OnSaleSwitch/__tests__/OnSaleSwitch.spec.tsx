import { cleanup, render, RenderResult } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import OnSaleSwitch from "..";

describe("<OnSaleSwitch />", () => {
  afterEach(cleanup);

  describe("when changing value", () => {
    let onSaleSwitch: RenderResult;
    let onChange: jest.Mock;
    beforeEach(async () => {
      onChange = jest.fn().mockName("onChange");
      onSaleSwitch = render(<OnSaleSwitch onChange={onChange} />);
    });

    it("calls the onChange handler with proper checked value", async () => {
      userEvent.click(onSaleSwitch.getByTestId("onSaleSwitch"));
      expect(onChange).toHaveBeenCalledWith(true);
    });
  });
});
