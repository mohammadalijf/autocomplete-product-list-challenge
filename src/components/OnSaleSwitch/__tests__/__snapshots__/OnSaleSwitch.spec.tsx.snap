// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OnSaleSwitch /> when changing value renders correctly 1`] = `
<DocumentFragment>
  <label
    class="MuiFormControlLabel-root"
  >
    <span
      class="MuiSwitch-root"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase MuiSwitch-colorPrimary"
        data-testid="onSaleSwitch"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            aria-label="show only on sales products"
            class="PrivateSwitchBase-input-4 MuiSwitch-input"
            type="checkbox"
            value=""
          />
          <span
            class="MuiSwitch-thumb"
          />
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiSwitch-track"
      />
    </span>
    <span
      class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      Show only on sales
    </span>
  </label>
</DocumentFragment>
`;
